src/prompt/functions/ai-chain.ts

```diff
@@ -0,0 +1,178 @@
+ *
+ * To prevent possible infinite loops, the chain will throw an error if it
+ * exceeds `maxCalls` (`maxCalls` is expected to be >= `maxRetries`).
+ *
+ * Note that `createAIChain` takes in a `functions` array of `AIFunction`
+ * objects created by `createAIFunction`, as the two utility functions are
+ * meant to be used together.
+ */
+export function createAIChain<
+  Params extends Prompt.ChainParams = void,
+  Result extends Prompt.ChainResult = string
+>(
+  {
+    chatModel,
+    schema,
+    params,
+    prompt,
+    functions,
+    maxCalls = 5,
+    maxRetries = 3,
+    toolCallConcurrency = 8,
+  }: {
+    chatModel: Model.Chat.Model;
+    schema?: ZodType<Result>;
+    params?: Omit<
+      Prettify<Model.Chat.Run & Model.Chat.Config>,
+      'functions' | 'tools' | 'function_call' | 'tool_choice'
+    >;
+    prompt?: Prompt.Template<Params>;
+    functions?: Prompt.AIFunction[];
+    maxCalls?: number;
+    maxRetries?: number;
+    toolCallConcurrency?: number;
+  },
+  context?: Model.Ctx
+): Prompt.Chain<Params, Result> {
+  return async (promptArgs: Params): Promise<Result> => {
+    if (!params?.messages && !prompt) {
+      throw new Error('Either params.messages or prompt must be provided');
+    }
+
+    if (params?.messages && prompt) {
+      throw new Error('Only one of params.messages or prompt may be provided');
+    }
+
+    const messages =
+      params?.messages ?? (await Promise.resolve(prompt!(promptArgs)));
+    const tools = functions?.map((func) => ({
+      type: 'function' as const,
+      function: func.spec,
+    }));
+    const functionMapByName = functions?.reduce<
+      Record<string, Prompt.AIFunction>
+    >(
+      (map, func) => ({
+        ...map,
+        [func.name]: func,
+      }),
+      {}
+    );
+
+    let numCalls = 0;
+    let numErrors = 0;
+
+    while (numCalls < maxCalls) {
+      ++numCalls;
+      const response = await chatModel.run(
+        {
+          ...params,
+          messages,
+          tools,
+        },
+        context
+      );
+
+      const { message } = response;
+      messages.push(message);
+
+      try {
+        if (Msg.isToolCall(message)) {
+          if (!functions) {
+            throw new AbortError('No functions provided to handle tool call');
+          }
+
+          await pMap(
+            message.tool_calls,
+            async (toolCall) => {
+              const func = functionMapByName![toolCall.function.name];
+
+              if (!func) {
+                throw new Error(
+                  `No function found with name ${toolCall.function.name}`
+                );
+              }
+
+              // TODO: ideally we'd differentiate between tool argument validation
+              // errors versus errors thrown from the tool implementation. Errors
+              // from the underlying tool could be things like network errors, which
+              // should be retried locally without re-calling the LLM.
```

Comment:
im in favor of adding a custom `ZodValidationError` that serves as both an indicator that it was a validation error and a way to narrow to the zod error type.